exonum/src/events/noise/mod.rs

```diff
@@ -40,80 +42,144 @@ pub struct HandshakeParams {
-        .and_then(move |(stream, msg)| {
-            let _buf = noise.read_handshake_msg(&msg);
-            write_handshake_msg(&mut noise)
+    fn write_handshake_msg<S: AsyncWrite + 'static>(
+        mut self,
+        stream: S,
+    ) -> Box<Future<Item = (S, Self), Error = io::Error>> {
+        Box::new(
+            done(self.noise.write_handshake_msg())
+                .map_err(|e| e.into())
                 .and_then(|(len, buf)| write(stream, &buf, len))
-                .and_then(move |(stream, _msg)| {
-                    let noise = noise.into_transport_mode()?;
-                    let framed = stream.framed(MessagesCodec::new(max_message_len, noise));
-                    Ok(framed)
-                })
-        });
+                .map(move |(stream, _)| (stream, self)),
+        )
+    }
+
+    fn finalize<S: AsyncRead + AsyncWrite + 'static>(
+        self,
+        stream: S,
+    ) -> Result<Framed<S, MessagesCodec>, io::Error> {
+        let noise = self.noise.into_transport_mode()?;
+        let framed = stream.framed(MessagesCodec::new(self.max_message_len, noise));
+        Ok(framed)
+    }
+}
+
+fn listen_handshake<S, T>(stream: S, handshake: T) -> HandshakeResult<S>
+where
+    S: AsyncRead + AsyncWrite + 'static,
+    T: Handshake + 'static,
+{
+    let framed = handshake
+        .read_handshake_msg(stream)
+        .and_then(|(stream, handshake)| handshake.write_handshake_msg(stream))
+        .and_then(|(stream, handshake)| handshake.read_handshake_msg(stream))
+        .and_then(|(stream, handshake)| handshake.finalize(stream));
+    Box::new(framed)
+}
 
+fn send_handshake<S, T>(stream: S, handshake: T) -> HandshakeResult<S>
+where
+    S: AsyncRead + AsyncWrite + 'static,
+    T: Handshake + 'static,
+{
+    let framed = handshake
+        .write_handshake_msg(stream)
+        .and_then(|(stream, handshake)| handshake.read_handshake_msg(stream))
+        .and_then(|(stream, handshake)| handshake.write_handshake_msg(stream))
+        .and_then(|(stream, handshake)| handshake.finalize(stream));
     Box::new(framed)
 }
 
-fn read(sock: TcpStream) -> Box<Future<Item = (TcpStream, Vec<u8>), Error = io::Error>> {
+fn read<S: AsyncRead + 'static>(sock: S) -> impl Future<Item = (S, Vec<u8>), Error = io::Error> {
     let buf = vec![0u8; HANDSHAKE_HEADER_LENGTH];
-    Box::new(
-        read_exact(sock, buf)
-            .and_then(|(stream, msg)| read_exact(stream, vec![0u8; msg[0] as usize])),
-    )
+    read_exact(sock, buf).and_then(|(stream, msg)| read_exact(stream, vec![0u8; msg[0] as usize]))
 }
 
-fn write(
-    sock: TcpStream,
+fn write<S: AsyncWrite + 'static>(
+    sock: S,
     buf: &[u8],
     len: usize,
-) -> Box<Future<Item = (TcpStream, Vec<u8>), Error = io::Error>> {
+) -> impl Future<Item = (S, Vec<u8>), Error = io::Error> {
     let mut message = vec![0u8; HANDSHAKE_HEADER_LENGTH];
+
+    if len > NOISE_MAX_HANDSHAKE_MESSAGE_LENGTH {
+        return Either::A(err(io::Error::new(
```

Comment:
I don't really like `Either` implementation with `A` and `B` branches. Can we somehow use [one](https://docs.rs/futures/0.2.1/futures/future/enum.Either.html) with `Left` and `Right`?

Reply:
👌right after we move to futures 0.2.